### ISSUE TITLE
freqai trade training throw 'DataFrame.dtypes for label must be int, float or bool' error

### ISSUE THREAD

**Issue by yvvw** [freqtrade/freqtrade#7988]
## Describe your environment
  * docker image: freqtradeorg/freqtrade:develop_freqai
  * Freqtrade Version: freqtrade docker-2023.1.dev-787d292b
  
## Your question

freqai training model throw warning '**freqtrade.freqai.freqai_interface - WARNING - Training ETH/USDT raised exception ValueError. Message: DataFrame.dtypes for label must be int, float or bool, skipping.**'

*command*
```shell
docker-compose run --rm freqtrade trade \
        --strategy FreqaiSg1Strategy \
        --freqaimodel LightGBMRegressor
```

*logs*
```log
2023-01-07 06:58:12,092 - freqtrade.freqai.base_models.BaseRegressionModel - INFO - -------------------- Starting training ETH/USDT --------------------
2023-01-07 06:58:12,250 - freqtrade.freqai.data_kitchen - INFO - dropped 0 training points due to NaNs in populated dataset 21599.
2023-01-07 06:58:12,251 - freqtrade.freqai.base_models.BaseRegressionModel - INFO - -------------------- Training on data from 2022-12-23 to 2023-01-07 --------------------
2023-01-07 06:58:12,369 - freqtrade.freqai.data_kitchen - INFO - SVM tossed 472 train points from 21599 total points.
2023-01-07 06:58:12,370 - freqtrade.freqai.data_kitchen - INFO - SVM tossed 472 test points from 21599 total points.
2023-01-07 06:58:13,198 - freqtrade.worker - INFO - Bot heartbeat. PID=1, version='docker-2023.1.dev-787d292b', state='RUNNING'
2023-01-07 06:58:17,920 - freqtrade.freqai.base_models.BaseRegressionModel - INFO - Training model on 336 features
2023-01-07 06:58:17,921 - freqtrade.freqai.base_models.BaseRegressionModel - INFO - Training model on 21127 data points
2023-01-07 06:58:17,928 - freqtrade.freqai.freqai_interface - WARNING - Training ETH/USDT raised exception ValueError. Message: DataFrame.dtypes for label must be int, float or bool, skipping.
2023-01-07 06:58:17,930 - freqtrade.freqai.freqai_interface - INFO - Total time spent training pairlist 6.21 seconds
```

*config.json* 
```json
{
    "bot_name": "freqai",
    "dry_run": true,
    "dry_run_wallet": 1000,
    "trading_mode": "futures",
    "margin_mode": "isolated",
    "fiat_display_currency": "USD",
    "stake_currency": "USDT",
    "stake_amount": 100,
    "max_open_trades": 10,
    "tradable_balance_ratio": 0.99,
    "timeframe": "1m",
    "cancel_open_orders_on_exit": true,
    "unfilledtimeout": {
        "entry": 10,
        "exit": 30
    },
    "freqai": {
        "enabled": true,
        "identifier": "sg1",
        "train_period_days": 15,
        "backtest_period_days": 7,
        "live_retrain_hours": 0.5,
        "purge_old_models": true,
        "continual_learning": true,
        "feature_parameters": {
            "include_timeframes": ["3m", "15m", "1h"],
            "include_corr_pairlist": ["BTC/USDT", "ETH/USDT"],
            "label_period_candles": 20,
            "include_shifted_candles": 2,
            "weight_factor": 0.9,
            "indicator_periods_candles": [10, 20],
            "principal_component_analysis": false,
            "plot_feature_importances": 0,
            "DI_threshold": 0.9,
            "use_SVM_to_remove_outliers": true
        },
        "data_split_parameters": {
            "test_size": 0,
            "random_state": 1
        },
        "model_training_parameters": {}
    },
    "exchange": {
        "name": "binance",
        "key": "",
        "secret": "",
        "ccxt_config": {},
        "ccxt_async_config": {},
        "pair_whitelist": [
            "ETH/USDT"
        ],
        "pair_blacklist": []
    },
    "force_entry_enable": true,
    "entry_pricing": {
        "price_side": "same",
        "use_order_book": true,
        "order_book_top": 1,
        "price_last_balance": 0.0,
        "check_depth_of_market": {
            "enabled": false,
            "bids_to_ask_delta": 1
        }
    },
    "exit_pricing": {
        "price_side": "other",
        "use_order_book": true,
        "order_book_top": 1
    },
    "pairlists": [
        {
            "method": "StaticPairList"
        }
    ],
    "initial_state": "running",
    "internals": {
        "process_throttle_secs": 5
    }
}

```

*strategy* [FreqaiExampleHybridStrategy.py](https://github.com/freqtrade/freqtrade/blob/develop/freqtrade/templates/FreqaiExampleHybridStrategy.py)

```python
import logging

import numpy as np
import pandas as pd
import talib.abstract as ta
from pandas import DataFrame
from technical import qtpylib

from freqtrade.strategy import IntParameter, IStrategy, merge_informative_pair


logger = logging.getLogger(__name__)


class FreqaiSg1Strategy(IStrategy):
    """
    Example of a hybrid FreqAI strat, designed to illustrate how a user may employ
    FreqAI to bolster a typical Freqtrade strategy.

    Launching this strategy would be:

    freqtrade trade --strategy FreqaiExampleHybridStrategy --strategy-path freqtrade/templates
    --freqaimodel CatboostClassifier --config config_examples/config_freqai.example.json

    or the user simply adds this to their config:

    "freqai": {
        "enabled": true,
        "purge_old_models": true,
        "train_period_days": 15,
        "identifier": "uniqe-id",
        "feature_parameters": {
            "include_timeframes": [
                "3m",
                "15m",
                "1h"
            ],
            "include_corr_pairlist": [
                "BTC/USDT",
                "ETH/USDT"
            ],
            "label_period_candles": 20,
            "include_shifted_candles": 2,
            "DI_threshold": 0.9,
            "weight_factor": 0.9,
            "principal_component_analysis": false,
            "use_SVM_to_remove_outliers": true,
            "indicator_periods_candles": [10, 20]
        },
        "data_split_parameters": {
            "test_size": 0,
            "random_state": 1
        },
        "model_training_parameters": {
            "n_estimators": 800
        }
    },

    Thanks to @smarmau and @johanvulgt for developing and sharing the strategy.
    """

    minimal_roi = {
        "60": 0.01,
        "30": 0.02,
        "0": 0.04
    }

    plot_config = {
        'main_plot': {
            'tema': {},
        },
        'subplots': {
            "MACD": {
                'macd': {'color': 'blue'},
                'macdsignal': {'color': 'orange'},
            },
            "RSI": {
                'rsi': {'color': 'red'},
            },
            "Up_or_down": {
                '&s-up_or_down': {'color': 'green'},
            }
        }
    }

    process_only_new_candles = True
    stoploss = -0.05
    use_exit_signal = True
    startup_candle_count: int = 30
    can_short = True

    # Hyperoptable parameters
    buy_rsi = IntParameter(low=1, high=50, default=30, space='buy', optimize=True, load=True)
    sell_rsi = IntParameter(low=50, high=100, default=70, space='sell', optimize=True, load=True)
    short_rsi = IntParameter(low=51, high=100, default=70, space='sell', optimize=True, load=True)
    exit_short_rsi = IntParameter(low=1, high=50, default=30, space='buy', optimize=True, load=True)

    # FreqAI required function, user can add or remove indicators, but general structure
    # must stay the same.
    def populate_any_indicators(
        self, pair, df, tf, informative=None, set_generalized_indicators=False
    ):
        """
        User feeds these indicators to FreqAI to train a classifier to decide
        if the market will go up or down.
        :param pair: pair to be used as informative
        :param df: strategy dataframe which will receive merges from informatives
        :param tf: timeframe of the dataframe which will modify the feature names
        :param informative: the dataframe associated with the informative pair
        """

        if informative is None:
            informative = self.dp.get_pair_dataframe(pair, tf)

        # first loop is automatically duplicating indicators for time periods
        for t in self.freqai_info["feature_parameters"]["indicator_periods_candles"]:

            t = int(t)
            informative[f"%-{pair}rsi-period_{t}"] = ta.RSI(informative, timeperiod=t)
            informative[f"%-{pair}mfi-period_{t}"] = ta.MFI(informative, timeperiod=t)
            informative[f"%-{pair}adx-period_{t}"] = ta.ADX(informative, timeperiod=t)
            informative[f"%-{pair}sma-period_{t}"] = ta.SMA(informative, timeperiod=t)
            informative[f"%-{pair}ema-period_{t}"] = ta.EMA(informative, timeperiod=t)
            informative[f"%-{pair}roc-period_{t}"] = ta.ROC(informative, timeperiod=t)
            informative[f"%-{pair}relative_volume-period_{t}"] = (
                informative["volume"] / informative["volume"].rolling(t).mean()
            )

        # FreqAI needs the following lines in order to detect features and automatically
        # expand upon them.
        indicators = [col for col in informative if col.startswith("%")]
        # This loop duplicates and shifts all indicators to add a sense of recency to data
        for n in range(self.freqai_info["feature_parameters"]["include_shifted_candles"] + 1):
            if n == 0:
                continue
            informative_shift = informative[indicators].shift(n)
            informative_shift = informative_shift.add_suffix("_shift-" + str(n))
            informative = pd.concat((informative, informative_shift), axis=1)

        df = merge_informative_pair(df, informative, self.config["timeframe"], tf, ffill=True)
        skip_columns = [
            (s + "_" + tf) for s in ["date", "open", "high", "low", "close", "volume"]
        ]
        df = df.drop(columns=skip_columns)

        # User can set the "target" here (in present case it is the
        # "up" or "down")
        if set_generalized_indicators:
            # User "looks into the future" here to figure out if the future
            # will be "up" or "down". This same column name is available to
            # the user
            df['&s-up_or_down'] = np.where(df["close"].shift(-50) >
                                           df["close"], 'up', 'down')

        return df

    # flake8: noqa: C901
    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        # User creates their own custom strat here. Present example is a supertrend
        # based strategy.

        dataframe = self.freqai.start(dataframe, metadata, self)

        # TA indicators to combine with the Freqai targets
        # RSI
        dataframe['rsi'] = ta.RSI(dataframe)

        # Bollinger Bands
        bollinger = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=2)
        dataframe['bb_lowerband'] = bollinger['lower']
        dataframe['bb_middleband'] = bollinger['mid']
        dataframe['bb_upperband'] = bollinger['upper']
        dataframe["bb_percent"] = (
            (dataframe["close"] - dataframe["bb_lowerband"]) /
            (dataframe["bb_upperband"] - dataframe["bb_lowerband"])
        )
        dataframe["bb_width"] = (
            (dataframe["bb_upperband"] - dataframe["bb_lowerband"]) / dataframe["bb_middleband"]
        )

        # TEMA - Triple Exponential Moving Average
        dataframe['tema'] = ta.TEMA(dataframe, timeperiod=9)

        return dataframe

    def populate_entry_trend(self, df: DataFrame, metadata: dict) -> DataFrame:

        df.loc[
            (
                # Signal: RSI crosses above 30
                (qtpylib.crossed_above(df['rsi'], self.buy_rsi.value)) &
                (df['tema'] <= df['bb_middleband']) &  # Guard: tema below BB middle
                (df['tema'] > df['tema'].shift(1)) &  # Guard: tema is raising
                (df['volume'] > 0) &  # Make sure Volume is not 0
                (df['do_predict'] == 1) &  # Make sure Freqai is confident in the prediction
                # Only enter trade if Freqai thinks the trend is in this direction
                (df['&s-up_or_down'] == 'up')
            ),
            'enter_long'] = 1

        df.loc[
            (
                # Signal: RSI crosses above 70
                (qtpylib.crossed_above(df['rsi'], self.short_rsi.value)) &
                (df['tema'] > df['bb_middleband']) &  # Guard: tema above BB middle
                (df['tema'] < df['tema'].shift(1)) &  # Guard: tema is falling
                (df['volume'] > 0) &  # Make sure Volume is not 0
                (df['do_predict'] == 1) &  # Make sure Freqai is confident in the prediction
                # Only enter trade if Freqai thinks the trend is in this direction
                (df['&s-up_or_down'] == 'down')
            ),
            'enter_short'] = 1

        return df

    def populate_exit_trend(self, df: DataFrame, metadata: dict) -> DataFrame:

        df.loc[
            (
                # Signal: RSI crosses above 70
                (qtpylib.crossed_above(df['rsi'], self.sell_rsi.value)) &
                (df['tema'] > df['bb_middleband']) &  # Guard: tema above BB middle
                (df['tema'] < df['tema'].shift(1)) &  # Guard: tema is falling
                (df['volume'] > 0)  # Make sure Volume is not 0
            ),

            'exit_long'] = 1

        df.loc[
            (
                # Signal: RSI crosses above 30
                (qtpylib.crossed_above(df['rsi'], self.exit_short_rsi.value)) &
                # Guard: tema below BB middle
                (df['tema'] <= df['bb_middleband']) &
                (df['tema'] > df['tema'].shift(1)) &  # Guard: tema is raising
                (df['volume'] > 0)  # Make sure Volume is not 0
            ),
            'exit_short'] = 1

        return df
```

**Comment by xmatthias** [comment]
If i'm honest, then i'd think that the message is quite clear.

```
2023-01-07 06:58:17,928 - freqtrade.freqai.freqai_interface - WARNING - Training ETH/USDT raised exception ValueError. Message: DataFrame.dtypes for label must be int, float or bool, skipping.
```

Your label definition is a string ('up', or 'down').
```
            df['&s-up_or_down'] = np.where(df["close"].shift(-50) > df["close"], 'up', 'down')
```

This is clearly originating from the Hybrid Strategy example (as you also point out above) - which is however made for classification models (`CatboostClassifier`, `LightGBMClassifier`, `XGBoostClassifier`, ...) - but not for regressors (which is what you used it with).

A regressor's target MUST have a numeric (or boolean) label - as it can't predict categories.
The example you posted also contains a runnable sample - which uses CatboostClassifier - so your simple fix is to use one of the Classifier models.

**Comment by yvvw** [comment]
thanks
